daemons/execd/remoted_tls.c

```diff
@@ -38,12 +42,190 @@ gnutls_dh_params_t dh_params;
+        free(path);
+        return;
+    }
+
+    /* The schema is a CDATA node, which is a child of the <file> node.  Traverse
+     * all children and look for the first CDATA child.  There can't be more than
+     * one because we only have one file attribute on the parent.
+     */
+    for (xmlNode *child = xml->children; child != NULL; child = child->next) {
+        if (child->type != XML_CDATA_SECTION_NODE) {
+            continue;
+        }
+
+        rc = pcmk__write_sync(fd, (const char *) child->content);
+        if (rc != pcmk_rc_ok) {
+            crm_err("Could not write schema file %s: %s", path, pcmk_rc_str(rc));
+        }
+
+        break;
+    }
+
+    free(path);
+    close(fd);
+}
+
+static void
+get_schema_files(void)
+{
+    int rc = pcmk_rc_ok;
+    cib_t *cib = NULL;
+    xmlNode *reply;
+
+    cib = cib_new();
+    if (cib == NULL) {
+        _exit(ENOTCONN);
+    }
+
+    rc = cib->cmds->signon(cib, crm_system_name, cib_query);
+    if (rc != pcmk_ok) {
+        crm_err("Could not connect to the CIB manager: %s", pcmk_strerror(rc));
+        _exit(pcmk_rc2exitc(rc));
+    }
+
+    rc = cib->cmds->fetch_schemas(cib, &reply, xml_latest_schema(), cib_sync_call);
+    if (rc != pcmk_ok) {
+        crm_err("Could not get schema files: %s", pcmk_strerror(rc));
+
+    } else {
+        bool need_to_load = false;
+
+        /* The returned document looks something like this:
+         * <cib_command>
+         *   <cib_calldata>
+         *     <schemas>
+         *       <schema version="pacemaker-1.1">
+         *         <file path="foo-1.1">
+         *         </file>
+         *         <file path="bar-1.1">
+         *         </file>
+         *         ...
+         *       </schema>
+         *       <schema version="pacemaker-1.2">
+         *       </schema>
+         *       ...
+         *     </schemas>
+         *   </cib_calldata>
+         * </cib_command>
+         *
+         * All the <schemas> and <schema> tags are really just there for organizing
+         * the XML a little better.  What we really care about are the <file> nodes,
+         * and specifically the path attributes and the CDATA children (not shown)
+         * of each.  We can use an xpath query to reach down and get all the <file>
+         * nodes at once.
+         *
+         * If we already have the latest schema version, or we asked for one later
+         * than what the cluster supports, we'll get back an empty <schemas> node,
+         * so all this will continue to work.  It just won't do anything.
+         */
+        xmlXPathObjectPtr xpathObj = xpath_search(reply, "//" PCMK__XA_FILE);
+
+        for (int i = 0; i < numXpathResults(xpathObj); i++) {
+            xmlNode *file_node = getXpathResult(xpathObj, i);
+
+            if (file_node == NULL) {
+                continue;
+            }
+
+            write_schema_file(file_node, CRM_REMOTE_SCHEMA_DIRECTORY);
+            need_to_load = true;
+        }
+
```

Comment:
Here's another function idea - a foreach function for xpath matches.

Reply:
Oh, this already exists - `crm_foreach_xpath_result`.  Guess I'll see if it's useful in this case.